src/uri.rs

```diff
@@ -389,6 +389,48 @@ impl StdError for UriError {
     }
 }
 
+// TODO: Is ByteStr adding value? Inline it?
```

Comment:
The `Bytes` type is sort of like an `Arc<Vec<u8>>`, which some improvements. The point is so that when a `Request` is parsed and constructed, the various pieces internally the are backed by a string of bytes, all point back to only a single shared buffer. So for a `Uri`, the string `"https://github.com/hyperium/hyper/pull/1226"` isn't copied into a new part of the heap (like it would if we used `String`), but its only one slot, and `Uri`'s `Bytes` keeps the indices of what part it's interested in.